### PR TITLE
Feature/migrate to androidx

### PR DESCRIPTION
**Pull request by CDRussell**
Task/Issue URL: https://app.asana.com/0/414730916066338/885014473436729
Tech Design URL: 
CC: 

**Description**:
Upgrade to `AndroidX`, and upgrade a bunch of dependencies along with it. Many dependencies weren't compatible with `AndroidX` at their existing version, so increasing their version was required to let our app upgrade.

Upgrading various libraries resulted in more stricter tooling checks so there are additional code tidy ups to fix many of these new errors/warnings.

**Steps to test this PR**:
1. Ensure the project builds (all dependencies resolved etc...)
1. Ensure all unit tests pass for you
1. App requires a **thorough** testing; there is a huge footprint of code changes here, and some new libraries to boot. So there's a lot of places where a regression could have crept in.

If possible, it would be good to get both @brindy and @subsymbolic to at least test the app out, even if only one of you will do the actual cod review. 🙏 

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: app/src/androidTest/java/com/duckduckgo/app/TestExtensions.kt [hunk 1]
```diff
@@ -16,14 +16,16 @@
...
+import androidx.annotation.UiThread
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.Observer
+import androidx.test.platform.app.InstrumentationRegistry
 import com.duckduckgo.app.di.TestAppComponent
 import java.util.concurrent.CountDownLatch
 import java.util.concurrent.TimeUnit
 
 // from https://stackoverflow.com/a/44991770/73479
+@UiThread
```
Comment (CDRussell) [inline app/src/androidTest/java/com/duckduckgo/app/TestExtensions.kt]: It is an error now to `observeForever` on a background thread

File: app/src/androidTest/java/com/duckduckgo/app/TestExtensions.kt [hunk 2]
```diff
@@ -37,7 +39,7 @@ fun <T> LiveData<T>.blockingObserve(): T? {
 }
 
 fun getApp(): TestApplication {
-    return InstrumentationRegistry.getTargetContext().applicationContext as TestApplication
+    return InstrumentationRegistry.getInstrumentation().targetContext.applicationContext as TestApplication
```
Comment (CDRussell) [inline app/src/androidTest/java/com/duckduckgo/app/TestExtensions.kt]: `InstrumentationRegistry.getInstrumentation().targetContext`

This is the new way to get hold of a `Context` for your target app during unit testing - so you'll see this change a lot in this PR. the old way is deprecated now - think i got them all.

File: app/src/androidTest/java/com/duckduckgo/app/browser/useragent/UserAgentProviderTest.kt [hunk 3]
```diff
@@ -24,10 +24,10 @@ private const val CHROME_UA_MOBILE =
 
 // Some values will be dynamic based on OS/Architecture/Software versions, so use Regex to match around dynamic values
 private val CHROME_UA_DESKTOP_REGEX = Regex(
-    "Mozilla/5.0 \\(X11; Linux .*?\\) AppleWebKit\\/[.0-9]+ \\(KHTML, like Gecko\\) Chrome\\/[.0-9]+ Safari/[.0-9]+"
+    "Mozilla/5.0 \\(X11; Linux .*?\\) AppleWebKit/[.0-9]+ \\(KHTML, like Gecko\\) Chrome/[.0-9]+ Safari/[.0-9]+"
```
Comment (CDRussell) [inline app/src/androidTest/java/com/duckduckgo/app/browser/useragent/UserAgentProviderTest.kt]: Lint warning that there was a needless escape char here

File: app/src/main/java/com/duckduckgo/app/global/DuckDuckGoActivity.kt [hunk 4]
```diff
@@ -56,7 +56,7 @@ abstract class DuckDuckGoActivity : AppCompatActivity() {
 
     override fun onDestroy() {
         themeChangeReceiver?.let {
-            LocalBroadcastManager.getInstance(applicationContext).unregisterReceiver(it)
+            androidx.localbroadcastmanager.content.LocalBroadcastManager.getInstance(applicationContext).unregisterReceiver(it)
```
Comment (subsymbolic) [inline app/src/main/java/com/duckduckgo/app/global/DuckDuckGoActivity.kt]: I don't see a package conflict. Can we remove `androidx.localbroadcastmanager.content`?

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/ColorChangingPageListener.kt [hunk 5]
```diff
@@ -16,13 +16,13 @@
...
-import android.support.annotation.ColorInt
-import android.support.v4.view.ViewPager
+import androidx.annotation.ColorInt
+import androidx.viewpager.widget.ViewPager
 import com.duckduckgo.app.global.view.ColorCombiner
 
 
 class ColorChangingPageListener(private val colorCombiner: ColorCombiner, private val newColorListener: NewColorListener) :
-    ViewPager.OnPageChangeListener {
+    androidx.viewpager.widget.ViewPager.OnPageChangeListener {
```
Comment (subsymbolic) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/ColorChangingPageListener.kt]: Can we remove the `androidx.viewpager.widget` package specifier from here?

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingPageFragment.kt [hunk 6]
```diff
@@ -37,7 +37,7 @@ import kotlinx.android.synthetic.main.content_onboarding_default_browser.*
 import timber.log.Timber
 import javax.inject.Inject
 
-sealed class OnboardingPageFragment : Fragment() {
+sealed class OnboardingPageFragment : androidx.fragment.app.Fragment() {
```
Comment (subsymbolic) [inline app/src/main/java/com/duckduckgo/app/onboarding/ui/OnboardingPageFragment.kt]: Remove package

File: app/src/main/java/com/duckduckgo/app/tabs/ui/TabSwitcherAdapter.kt [hunk 7]
```diff
@@ -94,7 +93,7 @@ class TabSwitcherAdapter(private val context: Context, private val itemClickList
         val url: TextView,
         val close: ImageView,
         val tabUnread: View
-    ) : RecyclerView.ViewHolder(root)
+    ) : androidx.recyclerview.widget.RecyclerView.ViewHolder(root)
```
Comment (subsymbolic) [inline app/src/main/java/com/duckduckgo/app/tabs/ui/TabSwitcherAdapter.kt]: Remove package

**Pull request merged**